frontend/src/pushapi.js

```diff
@@ -3,490 +3,432 @@
-        if (!pendingBindings.length) {
-          return;
-        }
-
-        var uniqueEventsList = [];
-        angular.forEach(pendingBindings, function (each) {
-          var eventKey = each.eventKey;
-
-          if (!(eventKey in boundEvents)) {
-            boundEvents[eventKey] = [];
-
-            // Only add events that haven't been subscribed for
-            uniqueEventsList.push(eventKey);
-          }
-          // Add handler binding
-          boundEvents[eventKey].push(each.handler);
-        });
-
-        // Clear list of pending bindings
-        pendingBindings.splice(0, pendingBindings.length);
-        subscribeForEvents(uniqueEventsList);
-      }
-
-      return service;
-    };
-  })
-
-  .config(function (pushApiProvider) {
-    var backendServiceUrl = '/pushService';
-    var keepAliveInterval = 30; //in seconds
-    var reconnectInterval = 10; //in seconds
-
-    pushApiProvider
-            .backendServiceUrl(backendServiceUrl)
-            .keepAliveTimeInterval(keepAliveInterval)
-            .reconnectTimeInterval(reconnectInterval);
-  })
-
-
-  /**
-   * @ngdoc directive
-   * @name pushHandler
-   * @restrict E
-   * 
-   * @description
-   * Binds push event handler and takes care of unbinding it when the directive is destroyed.
-   *
-   * @example
-   * As element:
-   * <push-handler event="MyPushEvent" correlation-id="vm.eventId" on-event="vm.handleEvent(data)"></push-handler>
-   *
-   * 
-   * @param {''} event - name of push event to bind handler to.
-   * @param {String} [correlationId] - correlationId for push event.
-   * @param {function(data)} onEvent - handler method to bind to push event.
-   */
-  .directive('pushHandler', function ($parse, pushApi) {
-    return {
-      restrict: 'E',
-      link: function (scope, elem, attrs) {
-        var eventName = attrs.event;
-        var correlationId = $parse(attrs.correlationId)(scope);
-        var onPushEvent = $parse(attrs.onEvent);
-
-        elem.remove();
-
-        var handler = pushApi.bindId(eventName, correlationId, function (eventData) {
-          onPushEvent(scope, {$event: eventData});
-        });
-
-        scope.$on('$destroy', function () {
-          pushApi.unbindId(eventName, correlationId, handler);
-        });
-      }
-    };
-  })
+   .provider('pushApi', function () {
+
+     this.endpoints = {
+       serviceUrl: ""
+     };
+
+     /**
+      * Sets the backend url connection.
+      *
+      * @param url the new url
+      * @returns {*} the provider instance for chaining purposes.
+      */
+     this.backendServiceUrl = function (url) {
+       this.endpoints.serviceUrl = url;
+       return this;
+     };
+
+     this.$get = function ($rootScope, $interval, $timeout, $window, $http, $q) {
+       var endpoints = this.endpoints;
+       var boundEvents = {};
+       var pendingBulkBindings = [];
+       var pendingInitialBindings = [];
+       var specialBindings = [];
```

Comment:
What is special about those special bindings?

Reply:
This is bindings with correlation ids. Maybe name should be ```bindingsWithCorrelationIds``` or ``` correlationIdBindings```